File: 637-average-of-levels-in-binary-tree/average-of-levels-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def averageOfLevels(self, root: Optional[TreeNode]) -> List[float]:
        if not root: return []
        
        res = []
        queue = deque([root])

        while len(queue) > 0:
            rowLength = len(queue)
            rowSum = 0

            for i in range(rowLength):
                curr = queue.popleft()
                rowSum += curr.val

                if curr.left: queue.append(curr.left)
                if curr.right: queue.append(curr.right)

            rowAverage = rowSum / rowLength
            res.append(rowAverage)
        
        return res

File: 637-average-of-levels-in-binary-tree/average-of-levels-in-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var averageOfLevels = function(root) {
    if (!root) return [];

    const res = [];
    const queue = [root];

    while (queue.length > 0) {
        const rowLength = queue.length;
        let rowSum = 0;

        for (let i = 0; i < rowLength; i++) {
            const curr = queue.shift();
            rowSum += curr.val;

            if(curr.left) queue.push(curr.left);
            if(curr.right) queue.push(curr.right);
        }   
        const rowAverage = rowSum / rowLength;
        res.push(rowAverage);
    }
    return res;
};